torch/cuda/__init__.py

```diff
@@ -322,40 +323,82 @@ def can_device_access_peer(device: _device_t, peer_device: _device_t) -> bool:
     return torch._C._cuda_canDeviceAccessPeer(device, peer_device)
 
 
-@contextlib.contextmanager
-def stream(stream):
+class StreamContext(object):
     r"""Context-manager that selects a given stream.
 
     All CUDA kernels queued within its context will be enqueued on a selected
     stream.
 
     Args:
+        Stream (Stream): selected stream. This manager is a no-op if it's
+            ``None``.
+    .. note:: Streams are per-device.
+    """
+    cur_stream : Optional['torch.cuda.Stream']
+
+    def __init__(self, stream: Optional['torch.cuda.Stream']):  # type: ignore
+        self.idx = -1
+        self.stream = stream
+        # Initialize the below streams to default stream on the current device
+        self.src_prev_stream = torch.cuda.default_stream(None)
+        self.dst_prev_stream = torch.cuda.default_stream(None)
+
+    def __enter__(self):
+        self.idx = _get_device_index(device=None, optional=True)
+        # If there is no CUDA device available, return
+        if self.idx == -1:
+            return
+
+        # Local cur_stream variable for type refinement
+        cur_stream = self.stream
```

Comment:
At least mypy is able to handle type refinement of class members. Small test case:

```
ezyang-mbp:~ ezyang$ cat a.py
from typing import Optional
class A:
    b: Optional[int]
    def f(self) -> None:
        assert self.b is not None
        reveal_type(self.b)
ezyang-mbp:~ ezyang$ mypy a.py
a.py:6: error: Revealed type is 'builtins.int'
```

Is TorchScript unable to do this?

Reply:
Yes, that's right.